Compiling with System.getProperty can throw NPE rather than useful error

<!--
 Please make sure your issue can be reproduced in the latest version of GWT.
 If possible, test in latest beta or release candidate (RC) as well.
-->

**GWT version:** 2.9, 2.10, 2.11
**Browser (with version):** n/a
**Operating System:** any

---

##### Description
Both binding properties and configuration properties (single value only) can be used to provide strings that can be read at compile time via `System.getProperty`, but if no value is set on a configuration property, a NPE will occur in the compiler.

Cases that should be supported:
* Multi-value configuration properties correctly emit an error:
    ```
       [ERROR] Errors in 'com/google/gwt/dev/jjs/test/SystemGetPropertyTest.java'
          [ERROR] Line 42: Property 'someConfigProperty' is multivalued. Multivalued properties are not supported by System.getProperty().
    ```
 * No property defined in .gwt.xml, either binding or configuration results in default value being used (if provided), otherwise error:
    ```
       [ERROR] Errors in 'com/google/gwt/dev/jjs/test/SystemGetPropertyTest.java'
          [ERROR] Line 45: Property 'unknownProperty' is not defined.
    ```
 * Configuration properties defined without a value however result in an NPE:
    ```
       Compiling 1 permutation
          [ERROR] An internal compiler exception occurred
    com.google.gwt.dev.jjs.InternalCompilerException: Unexpected error during visit.
        at com.google.gwt.dev.jjs.ast.JVisitor.translateException(JVisitor.java:111)
        at com.google.gwt.dev.jjs.ast.JModVisitor.acceptImmutable(JModVisitor.java:313)
        at com.google.gwt.dev.jjs.ast.JMethodCall.visitChildren(JMethodCall.java:275)
        at com.google.gwt.dev.jjs.ast.JMethodCall.traverse(JMethodCall.java:266)
        at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
        at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:118)
        at com.google.gwt.dev.jjs.ast.JExpressionStatement.traverse(JExpressionStatement.java:42)
        at com.google.gwt.dev.jjs.ast.JModVisitor$ListContext.traverse(JModVisitor.java:88)
        at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemove(JModVisitor.java:331)
        at com.google.gwt.dev.jjs.ast.JBlock.traverse(JBlock.java:94)
        at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
        at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:139)
        at com.google.gwt.dev.jjs.ast.JVisitor.accept(JVisitor.java:135)
        at com.google.gwt.dev.jjs.ast.JMethodBody.traverse(JMethodBody.java:83)
        at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
        at com.google.gwt.dev.jjs.ast.JMethod.visitChildren(JMethod.java:786)
        at com.google.gwt.dev.jjs.ast.JMethod.traverse(JMethod.java:778)
        at com.google.gwt.dev.jjs.ast.JModVisitor$ListContextImmutable.traverse(JModVisitor.java:169)
        at com.google.gwt.dev.jjs.ast.JModVisitor.acceptWithInsertRemoveImmutable(JModVisitor.java:336)
        at com.google.gwt.dev.jjs.ast.JClassType.traverse(JClassType.java:147)
        at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
        at com.google.gwt.dev.jjs.ast.JProgram.visitModuleTypes(JProgram.java:1284)
        at com.google.gwt.dev.jjs.ast.JProgram.traverse(JProgram.java:1249)
        at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:273)
        at com.google.gwt.dev.jjs.ast.JModVisitor.accept(JModVisitor.java:265)
        at com.google.gwt.dev.jjs.impl.ResolvePermutationDependentValues.execImpl(ResolvePermutationDependentValues.java:163)
        at com.google.gwt.dev.jjs.impl.ResolvePermutationDependentValues.exec(ResolvePermutationDependentValues.java:113)
        at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:333)
        at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:272)
        at com.google.gwt.dev.CompilePerms.compile(CompilePerms.java:198)
        at com.google.gwt.dev.ThreadedPermutationWorkerFactory$ThreadedPermutationWorker.compile(ThreadedPermutationWorkerFactory.java:50)
        at com.google.gwt.dev.PermutationWorkerFactory$Manager$WorkerThread.run(PermutationWorkerFactory.java:74)
        at java.base/java.lang.Thread.run(Thread.java:829)
    Caused by: java.lang.NullPointerException
        at com.google.gwt.thirdparty.guava.common.base.Preconditions.checkNotNull(Preconditions.java:212)
        at com.google.gwt.thirdparty.guava.common.base.Joiner.toString(Joiner.java:447)
        at com.google.gwt.thirdparty.guava.common.base.Joiner.appendTo(Joiner.java:110)
        at com.google.gwt.thirdparty.guava.common.base.Joiner.appendTo(Joiner.java:154)
        at com.google.gwt.thirdparty.guava.common.base.Joiner.join(Joiner.java:197)
        at com.google.gwt.thirdparty.guava.common.base.Joiner.join(Joiner.java:186)
        at com.google.gwt.dev.jjs.impl.ResolvePermutationDependentValues$ValueReplacer.propertyValueExpression(ResolvePermutationDependentValues.java:79)
        at com.google.gwt.dev.jjs.impl.ResolvePermutationDependentValues$ValueReplacer.endVisit(ResolvePermutationDependentValues.java:73)
        at com.google.gwt.dev.jjs.ast.JPermutationDependentValue.traverse(JPermutationDependentValue.java:112)
        at com.google.gwt.dev.jjs.ast.JModVisitor.traverse(JModVisitor.java:361)
        at com.google.gwt.dev.jjs.ast.JModVisitor.acceptImmutable(JModVisitor.java:305)
        ... 39 more
             [ERROR] at SystemGetPropertyTest.java(44): System.getProperty("someConfigPropertyUnset", "bar")
             [ERROR] at SystemGetPropertyTest.java(44): Assert.assertEquals("bar", System.getProperty("someConfigPropertyUnset", "bar"))
             [ERROR] at SystemGetPropertyTest.java(44): Assert.assertEquals("bar", System.getProperty("someConfigPropertyUnset", "bar"))
             [ERROR] at SystemGetPropertyTest.java(41): {
      Assert.assertEquals("foo", "foo");
      Assert.assertEquals("foo", "foo");
      Assert.assertEquals("bar", System.getProperty("someConfigPropertyUnset", "bar"));
    }
             [ERROR] at SystemGetPropertyTest.java(41): {
      Assert.assertEquals("foo", "foo");
      Assert.assertEquals("foo", "foo");
      Assert.assertEquals("bar", System.getProperty("someConfigPropertyUnset", "bar"));
    }
             [ERROR] at SystemGetPropertyTest.java(41): com.google.gwt.dev.jjs.test.SystemGetPropertyTest.testConfigurationProperties()V
             [ERROR] at SystemGetPropertyTest.java(23): com.google.gwt.dev.jjs.test.SystemGetPropertyTest (extends GWTTestCase)
             [ERROR] at Unknown(0): <JProgram>
          [ERROR] Unrecoverable exception, shutting down
    ```


##### Steps to reproduce
Declare a configuration property, but don't assign it (for example, one .gwt.xml for a library might declare it, and a downstream app may choose to set it):
```xml
  <define-configuration-property name="someConfigPropertyUnset" is-multi-valued="false"/>
```
Attempt to use that property in GWT/Java sources without a default:
```java
    assertEquals("bar", System.getProperty("someConfigPropertyUnset", "bar"));
```
Expected: default value is used
Actual: NPE is thrown during compilation

##### Known workarounds
The simplest workaround is to just define a "this is unset" value for the configuration property, but that means that shared code has two places to change the default.
```xml
  <set-configuration-property name="someConfigPropertyUnset" value="this-is-unset" />
```

--- a/dev/core/src/com/google/gwt/dev/cfg/ConfigurationProperties.java
+++ b/dev/core/src/com/google/gwt/dev/cfg/ConfigurationProperties.java
@@ -142,6 +142,17 @@ public class ConfigurationProperties implements Serializable {
   }
 
   /**
+   * Returns all the values of a multi-valued configuration property, or null
+   * if not found.
+   *
+   * <p>A single-valued and unset configuration property will be returned as a list
+   * containing one null.
+   */
+  public List<String> getStringsOrNull(String key) {
+     return properties.get(key);
+  }
+
+  /**
    * Reads a configuration property as a comma-separated list of strings.
    * It may be a single-valued or multi-valued property. If multi-valued,
    * each value will be split on commas and all items concatenated.

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/ResolvePermutationDependentValues.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/ResolvePermutationDependentValues.java
@@ -47,6 +47,7 @@ import com.google.gwt.thirdparty.guava.common.collect.Multimap;
 import java.util.Collection;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 
 /**
@@ -76,7 +77,14 @@ public class ResolvePermutationDependentValues {
     private JExpression propertyValueExpression(JPermutationDependentValue x) {
       List<String> propertyValues = props.getConfigurationProperties().getStrings(x.getRequestedValue());
 
-      String propertyValue = propertyValues.isEmpty() ? null : Joiner.on(",").join(propertyValues);
+      if (!propertyValues.isEmpty() && propertyValues.stream().allMatch(Objects::isNull)) {
+        if (x.getDefaultValueExpression() != null) {
+          return x.getDefaultValueExpression();
+        }
+        throw new InternalCompilerException("No default set for configuration property '"
+            + x.getRequestedValue() + "'");
+      }
+      String propertyValue = propertyValues.isEmpty() ? null : Joiner.on(",").skipNulls().join(propertyValues);
 
       if (propertyValue != null) {
         // It is a configuration property.

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/ResolvePermutationDependentValues.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/ResolvePermutationDependentValues.java
@@ -75,20 +75,20 @@ public class ResolvePermutationDependentValues {
     }
 
     private JExpression propertyValueExpression(JPermutationDependentValue x) {
-      List<String> propertyValues = props.getConfigurationProperties().getStrings(x.getRequestedValue());
-
-      if (!propertyValues.isEmpty() && propertyValues.stream().allMatch(Objects::isNull)) {
+      List<String> propertyValues = props.getConfigurationProperties()
+          .getStringsOrNull(x.getRequestedValue());
+      if (propertyValues != null) {
+        // It is a configuration property.
+        // If no values are set, propertyValues is either empty (multivalued properties)
+        // or contains a single null (other properties).
+        if (propertyValues.stream().anyMatch(Objects::nonNull)) {
+          return program.getLiteral(x.getSourceInfo(),
+              Joiner.on(",").skipNulls().join(propertyValues));
+        }
         if (x.getDefaultValueExpression() != null) {
           return x.getDefaultValueExpression();
         }
-        throw new InternalCompilerException("No default set for configuration property '"
-            + x.getRequestedValue() + "'");
-      }
-      String propertyValue = propertyValues.isEmpty() ? null : Joiner.on(",").skipNulls().join(propertyValues);
-
-      if (propertyValue != null) {
-        // It is a configuration property.
-        return program.getLiteral(x.getSourceInfo(), propertyValue);
+        return program.getLiteralNull();
       }
 
       if (isSoftPermutationProperty(x.getRequestedValue())) {
@@ -96,7 +96,7 @@ public class ResolvePermutationDependentValues {
         return new JMethodCall(x.getSourceInfo(), null, method);
       }
 
-      propertyValue = commonPropertyAndBindingInfo.getPropertyValue(x.getRequestedValue());
+      String propertyValue = commonPropertyAndBindingInfo.getPropertyValue(x.getRequestedValue());
 
       if (propertyValue != null) {
         return program.getLiteral(x.getSourceInfo(), propertyValue);

--- a/user/test/com/google/gwt/dev/jjs/SystemGetPropertyTest.gwt.xml
+++ b/user/test/com/google/gwt/dev/jjs/SystemGetPropertyTest.gwt.xml
@@ -33,4 +33,5 @@
   <set-property name="someOtherDynamicProperty" value="blue">
     <when-property-is name="collapsedProperty" value="two"/>
   </set-property>
+  <define-configuration-property name="configPropertyUnset" is-multi-valued="false"/>
 </module>

--- a/user/test/com/google/gwt/dev/jjs/SystemGetPropertyTest.gwt.xml
+++ b/user/test/com/google/gwt/dev/jjs/SystemGetPropertyTest.gwt.xml
@@ -34,4 +34,5 @@
     <when-property-is name="collapsedProperty" value="two"/>
   </set-property>
   <define-configuration-property name="configPropertyUnset" is-multi-valued="false"/>
+  <define-configuration-property name="multivaluedConfigPropertyUnset" is-multi-valued="true"/>
 </module>

--- a/user/test/com/google/gwt/dev/jjs/test/SystemGetPropertyTest.java
+++ b/user/test/com/google/gwt/dev/jjs/test/SystemGetPropertyTest.java
@@ -37,5 +37,8 @@ public class SystemGetPropertyTest extends GWTTestCase {
         "InSafari" : "NotInSafari";
     assertEquals(expectedResult, System.getProperty("someDynamicProperty"));
     assertEquals("foo", System.getProperty("configPropertyUnset", "foo"));
+    assertNull(System.getProperty("configPropertyUnset"));
+    assertEquals("foo", System.getProperty("multivaluedConfigPropertyUnset", "foo"));
+    assertNull(System.getProperty("multivaluedConfigPropertyUnset"));
   }
 }

--- a/user/test/com/google/gwt/dev/jjs/test/SystemGetPropertyTest.java
+++ b/user/test/com/google/gwt/dev/jjs/test/SystemGetPropertyTest.java
@@ -36,5 +36,6 @@ public class SystemGetPropertyTest extends GWTTestCase {
     String expectedResult = "safari".equals(System.getProperty("user.agent")) ?
         "InSafari" : "NotInSafari";
     assertEquals(expectedResult, System.getProperty("someDynamicProperty"));
+    assertEquals("foo", System.getProperty("configPropertyUnset", "foo"));
   }
 }

